text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,343 @@
+selectively choose which Rust crates will be instrumented and have tests run. Since potentially unit tests from
+multiple crates will be run, the code coverage results will include coverage results for a crate that is being
+covered by a test from another crate.
+
+The `--coverage` flag will also support a comma separated list of crates so that a Rust developer can control
+the exact set of crates which will be instrumented during a build and test run.
+
+For example, let's take the following workspace with default members:
+
+```text
+/Cargo.toml
+/member1
+  +-- src
+    +-- lib.rs
+/member2
+  +-- src
+    +-- lib.rs
+/member3
+  +-- src
+    +-- lib.rs
+```
+
+Where crate `foo` has a virtual manifest:
+```toml
+[workspace]
+members = ["path/to/member1", "path/to/member2", "path/to/member3"]
+default-members = ["path/to/member2"]
+```
+
+Running the following cargo invocation would only instrument the workspace member
+`member2` and run the tests for `member2`:
+
+```
+cargo test --coverage
+```
+
+To instrument a different workspace member, the following command is also supported:
+
+```
+cargo test --coverage=member1,member3
+```
+
+This cargo invocation will instruct Cargo to run the tests for the default workspace member `member2`, but
+will only instrument the crates `member1` and `member3` to collect coverage results for. This flexibility
+allows a Rust developer to see just how much a given crate is covered by tests from another crate.
+
+With this feature, existing custom Cargo commands can leverage Cargo to do all of the heavy work to instrument specific
+crates and generate coverage results in a multitude of formats. This will allow in more flexibilty in generating
+for custom commands that generate code coverage for crates today.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As mentioned earlier, this RFC proposes adding multiple flags to the `cargo test` command. The flag `--coverage`
+would be responsible for setting the `-C instrument-coverage` flag that Cargo would pass on to the rustc invocation of the
+current crate. In the previous example, `foo` would be the crate being instrumented and `regex` would be an upstream dependency.
+
+Using the `--coverage` flag, Cargo would only set the `-C instrument-coverage` flag for the crate `foo`. If the `RUSTFLAGS`
+environment variable had already been set to include the `-C instrument-coverage` flag, then Cargo would still pass that
+flag to all crates within the dependency graph, including the `regex` crate and any transitive dependencies.
+
+This should not break any existing workflows and is strictly an opt-in feature.
+
+To use this new feature do the following:
+
+`cargo test --coverage`
+
+This flag would also be responsible for setting the `LLVM_PROFILE_FILE` environment variable which is read by LLVM
+to generate a `.profraw` file for each test executable that is run. Once again if the environment variable is already set,
```

Comment:
> file for each test executable

What precautions do we need to take to ensure uniqueness if two crates in the workspace have the same test executable name?

Reply:
> > file for each test executable
> 
> What precautions do we need to take to ensure uniqueness if two crates in the workspace have the same test executable name?

As mentioned in an above comment, the LLVM accepts special parameters for the `LLVM_PROFILE_FILE` environment variable. These parameters allow for ensuring uniqueness across all test runs by using a combination of the process ID and binary signature.

Reply:
llvm puts the profraw in the folder of the workspace crate the test was defined in - not the workspace root. But adding pid or timestamp to the `LLVM_PROFILE_FILE` would be _extra safe_